src/iteration.rs

```diff
@@ -0,0 +1,295 @@
+
+struct HeapReadSeekerPipe<R: Read + Seek> {
+    reader: R,
+    buffer: [u8; READER_BUFFER_SIZE],
+}
+
+/// The contents of a file, yielded in order from the beginning to the end of
+/// the archive.
+///
+/// Each file, including directories, will have a StartOfFile, zero or more
+/// chunks of data, and then a corresponding EndOfFile to mark that the file is
+/// done.
+pub enum ArchiveContents {
+    /// Marks the start of a file, including for directories.
+    StartOfFile(String),
+    /// A chunk of uncompressed data from the file. Files may have zero or more
+    /// chunks.
+    DataChunk(Vec<u8>),
+    /// Marks the end of the file that was started by the previous StartOfFile.
+    EndOfFile,
+    Err(Error),
+}
+
+/// An iterator over the contents of an archive.
+pub struct ArchiveIterator<R: Read + Seek> {
+    archive_entry: *mut ffi::archive_entry,
+    archive_reader: *mut ffi::archive,
+
+    in_file: bool,
+    closed: bool,
+    error: bool,
+
+    _pipe: Box<HeapReadSeekerPipe<R>>,
+    _utf8_guard: UTF8LocaleGuard,
+}
+
+impl<R: Read + Seek> Iterator for ArchiveIterator<R> {
+    type Item = ArchiveContents;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.error {
+            return None;
+        }
+
+        let next = if self.in_file {
+            unsafe { self.next_data_chunk() }
+        } else {
+            unsafe { self.next_header() }
+        };
+
+        match &next {
+            ArchiveContents::StartOfFile(_) => {
+                self.in_file = true;
+                Some(next)
+            }
+            ArchiveContents::DataChunk(_) => Some(next),
+            ArchiveContents::EndOfFile if self.in_file => {
+                self.in_file = false;
+                Some(next)
+            }
+            ArchiveContents::EndOfFile => None,
+            ArchiveContents::Err(_) => {
+                self.error = true;
+                Some(next)
+            }
+        }
+    }
+}
+
+impl<R: Read + Seek> Drop for ArchiveIterator<R> {
+    fn drop(&mut self) {
+        if !self.closed {
+            unsafe {
+                ffi::archive_read_close(self.archive_reader);
+                ffi::archive_read_free(self.archive_reader);
+            }
+        }
+    }
+}
+
+impl<R: Read + Seek> ArchiveIterator<R> {
+    /// Close the iterator, freeing up the associated resources.
+    ///
+    /// Resources will be freed on drop if this is not called, but any errors
+    /// during freeing on drop will be lost.
+    pub fn close(self) -> Result<()> {
+        unsafe {
+            archive_result(
+                ffi::archive_read_close(self.archive_reader),
+                self.archive_reader,
+            )?;
+            archive_result(
+                ffi::archive_read_free(self.archive_reader),
+                self.archive_reader,
+            )?;
+        }
+        let mut iter = self;
+        iter.closed = true;
+        Ok(())
```

Comment:
This could reuse the drop method.

Reply:
I can but that would mean allocating room for the result of the frees and retrieving it. I could make this method take a `&mut self` and use this method on drop, but then the ergonomics change since close() no longer consumes the value. I considered refactoring it out into a common method but that only saves two simple lines of code and wastes a few cycles constructing result objects to be discarded.